File: Loadtests/SignalRLoad/SignalRLoad/Hubs/LoadHub.cs
﻿using System;
using Microsoft.AspNet.SignalR;
using Shared.Extensions;
using Shared.Models;
using Shared.Utils;
using SignalRLoad.Models;

namespace SignalRLoad.Hubs
{
    public class LoadHub : Hub
    {
        private readonly Monitor _monitor;

        public LoadHub()
        {
            _monitor = Monitor.GetInstance();
        }

        public void InitTest(string testToRun, int numberOfClients, int spacing, long startTime)
        {
            _monitor.Reset();
            _monitor.NumberOfClients = numberOfClients;
            _monitor.Spacing = spacing;

            _monitor.ClientStartTime = DateUtils.FromMillisecondsSinceEpoch(startTime); //One hour time difference from client for some reason
            _monitor.ServerStartTime = DateTime.UtcNow;
            Clients.All.initTest(testToRun);
        }

        public void Echo(Message message)
        {
            RegisterReceivedAndSentFromClientEvents(message);
            Clients.Caller.receiveMessage(message);
            RegisterSentFromServerEvent(false);
        }

        public void Broadcast(Message message)
        {
            RegisterReceivedAndSentFromClientEvents(message);
            Clients.All.receiveMessage(message);
            RegisterSentFromServerEvent(true);
        }

        public void Complete(string clientId)
        {
            _monitor.CompletedClients.Add(clientId);

            if (!_monitor.Complete()) return;

            _monitor.Duration = DateTime.UtcNow.ToMilliseconds() - _monitor.ServerStartTime.ToMilliseconds();
            Clients.All.harvest();
        }

        public void GetData(TestDataEntity testData, int numberOfClientsInBrowser)
        {
            _monitor.TestDataEntities.Add(testData);
            _monitor.Harvested += numberOfClientsInBrowser;

            if (_monitor.HarvestedAll())
            {
                Clients.All.harvestComplete(new
                {
                    Duration = _monitor.Duration,
                    StartTime = _monitor.ClientStartTime.ToMilliseconds(),
                    SentFromClientEvents = _monitor.SentFromClientEvents,
                    ReceivedAtServerEvents = _monitor.ReceivedAtServerEvents,
                    SentFromServerEvents = _monitor.SentFromServerEvents,
                    Spacing = _monitor.Spacing,
                    TestDataEntities = _monitor.TestDataEntities
                });
            }
        }

        private void RegisterReceivedAndSentFromClientEvents(Message message)
        {
            message.ReceivedAtServer = DateTime.UtcNow.ToMilliseconds();    //One hour time difference from client for some reason                   
            _monitor.RegisterReceivedAtServerEvent(message.ReceivedAtServer, _monitor.Spacing);
            var key = _monitor.RegisterSentFromClientEvent(message.SentFromClient, _monitor.Spacing);
            message.Key = key;
        }

        private void RegisterSentFromServerEvent(bool broadCast)
        {
            var sent = DateTime.UtcNow.ToMilliseconds();
            _monitor.RegisterSentFromServerEvent(sent, broadCast, _monitor.Spacing);
        }
    }
}

File: Loadtests/PlayLoad/app/hubs/LoadHub.java
package hubs;

import java.util.Calendar;
import java.util.concurrent.ConcurrentHashMap;

import models.*;

public class LoadHub { 
	private Monitor _monitor;
	public ConcurrentHashMap<String, Socket> members;
	
	public LoadHub() {
		_monitor = Monitor.getInstance();
		members = new ConcurrentHashMap<>();
	}
	
	public void initTest(String testTorRun, int numberOfClients, int spacing, long startTime) {
		_monitor.reset();
        _monitor.numberOfClients = numberOfClients;
        _monitor.spacing = spacing;

        _monitor.clientStartTime = startTime;
        _monitor.serverStartTime = Calendar.getInstance().getTimeInMillis();
	}

	public void echo(Message message) {
		registerReceivedAndSentFromClientEvents(message);
        registerSentFromServerEvent(false);		
	}

	public void broadcast(Message message) {
		registerReceivedAndSentFromClientEvents(message);
        registerSentFromServerEvent(true);		
	}

	public boolean complete(String clientId) {
		 _monitor.completedClients.add(clientId);

        if (!_monitor.complete()) return false;

        _monitor.duration = Calendar.getInstance().getTimeInMillis() - _monitor.serverStartTime;
        return true;
	}

	public boolean getData(TestDataEntity testData, int numberOfClientsInBrowser) {
		 _monitor.testDataEntities.add(testData);
         _monitor.harvested += numberOfClientsInBrowser;
         
         return _monitor.harvestedAll();
	} 
	
	private void registerReceivedAndSentFromClientEvents(Message message)
    {
        message.ReceivedAtServer = Calendar.getInstance().getTimeInMillis();                   
        _monitor.registerReceivedAtServerEvent(message.ReceivedAtServer, _monitor.spacing);
        int key = _monitor.registerSentFromClientEvent(message.SentFromClient, _monitor.spacing);
        message.Key = key;
    }

    private void registerSentFromServerEvent(boolean broadCast)
    {
        long sent = Calendar.getInstance().getTimeInMillis();
        _monitor.registerSentFromServerEvent(sent, broadCast, _monitor.spacing);
    }

}
